Add auto-configuration for `spring-integration-http`

artembilan opened the pull request:
* Bump SI version to `4.3.0.RELEASE`
* Add `IntegrationProperties` `@ConfigurationProperties` based on the `/META-INF/spring.integration.default.properties`
* Register `globalIntegrationProperties` based on the `IntegrationProperties`, like it is required by Spring Integration, if any
* Register `@EnableIntegrationGraphController` for the `Controller` to expose integration runtime graph as a REST Service

wilkinsona commented on spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/integration/IntegrationAutoConfiguration.java:
```diff
@@ -39,19 +47,49 @@
+	@ConditionalOnWebApplication
+	@ConditionalOnClass(EnableIntegrationGraphController.class)
+	@ConditionalOnMissingBean(value = IntegrationGraphController.class, search = SearchStrategy.CURRENT)
+	@EnableIntegrationGraphController(path = "${spring.integration.graph-controller-path}")
```
How is the `path` value handled? I don't think this will support relaxed binding, e.g. if someone sets `spring.integration.graphControllerPath` or `SPRING_INTEGRATION_GRAPH_CONTROLLER_PATH` the configuration will be ignored.

snicoll replied:
+1 - And the use of such expression for the JMX stuff always bugged me.

wilkinsona commented:
The actual version upgrade is a duplicate of #5079 which has already been fixed

artembilan commented:
@wilkinsona , @snicoll , I addressed your concerns and rebased to the `master`.
Will hope that I have guessed what you mean :smile: 

Thank you for taking a look though :smile:

snicoll commented:
There are actually quite a lot of new things there that are not 4.3 specific. @wilkinsona and I had a more deep look to the code and we have some concerns. For one I'd prefer to have separate PRs instead of mixing several things into a single PR.

@artembilan could you please submit several PRs rather than the current one? Before you do we have several remarks that we're going to add as comments.

snicoll commented on spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/integration/IntegrationProperties.java:
```diff
@@ -0,0 +1,140 @@
+	private final Channels channels = new Channels();
+
+	/**
+	 * The number of threads available in the default taskScheduler.
```
Configuration key description do not start with "The". See the appendix for examples.

artembilan replied:
Thank you all for all the feedback!
I'll try to follow with your requirements and will go ahead with short steps.

Please, join me to that discussion about default `TaskScheduler`. We definitely should consider to reuse that auto-configured for Spring Integration as well. Although that might be just like copy/paste options because our default is like : 
```
BeanDefinition scheduler = BeanDefinitionBuilder.genericBeanDefinition(ThreadPoolTaskScheduler.class)
		.addPropertyValue("poolSize", IntegrationProperties.getExpressionFor(IntegrationProperties.TASK_SCHEDULER_POOL_SIZE))
		.addPropertyValue("threadNamePrefix", "task-scheduler-")
		.addPropertyValue("rejectedExecutionHandler", new CallerRunsPolicy())
		.addPropertyValue("errorHandler", new RootBeanDefinition(MessagePublishingErrorHandler.class))
		.getBeanDefinition();
```

The most important option is `MessagePublishingErrorHandler`.

OK. Never mind. We can discuss that on its own turn.

snicoll commented on spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/integration/IntegrationProperties.java:
```diff
@@ -0,0 +1,140 @@
+	private int taskSchedulerPoolSize = 10;
+
+	/**
+	 * When {@code true}, messages that arrive at a gateway reply channel will throw an exception.
```
Configuration key description shouldn't use Javadoc tag. There are many other use in this class.

snicoll commented on spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/integration/IntegrationProperties.java:
```diff
@@ -0,0 +1,140 @@
+	/**
+	 * When {@code true}, Messaging Annotation Support requires a declaration of the {@code @MessageEndpoint}.
+	 */
+	private boolean componentAnnotationRequired = false;
```
Why would you want to expose that as a property?

artembilan replied:
No-op. That was @garyrussell 's wish to avoid some breaking changes: https://github.com/spring-projects/spring-integration/pull/1760#issuecomment-208504119 :smile: 

We can reconsider that in SI-5.0 and yes, remove altogether and don't expose it as an auto-config option at all.

snicoll commented on spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/integration/IntegrationProperties.java:
```diff
@@ -0,0 +1,140 @@
+		private boolean autoCreate = true;
+
+		/**
+		 * This property provides the default number of subscribers allowed on, say, a {@code DirectChannel}.
```
"This property" isn't a valid prefix for a description.

wilkinsona commented:
@artembilan The scope of this PR seems to have grown quite a lot (or perhaps I missed something on my first review) since you originally opened it. It's doing a lot more than upgrading to SI 4.3. @snicoll's new title doesn't even cover everything.

I can see several separate changes:

 1. Fix the use of `spring.jmx.default-domain` so that relaxed binding will work
 2. Allow Spring Integration's global properties to be configured via the environment
 3. Auto-configuration for the new graph controller

Can you please split these up into separate pull requests? A few comments about the changes:

1. This would be a good candidate for inclusion in 1.3.x. 
2. AFAICT, most of these properties have been around since before Spring Integration 4.3 so they don't seem to be tied to an upgrade to 4.3. What's the motivation for now adding support for configuring them via the environment?
3. Using a separate property source to configure the path feels like a hack. I wonder if it would be possible to programatically create the Spring MVC mapping instead?

Hopefully it'll be easier to discuss and refine the changes once they're in separate pull requests. Just to set expectations; we're running out of time for 1.4 (RC1 is a week today) so, other than the first change (which is really a bug fix), these changes may not make it into 1.4.

artembilan commented:
@wilkinsona , fair enough!

I'll PR you for JMX fix and come back for all other concerns in your `1.5` dead line.

I would like to know too, if it is really possible to configure `@RequestMapping` programmatically...

OK, I'll dig about that eventually.
